### PR TITLE
Ockam node healthcheck application.

### PR DESCRIPTION
**Pull request by hairyhum**
This application periodically calls configured endpoint by connecting
via secure channel and sending a message to the healthcheck worker.
If message comes back, healthcheck application reports `ockam_healthcheck_ok_count` metric,
otherwise `ockam_healthcheck_errors_count` metric.

Configuration:
CHECK_NODE_HOST - hostname for checked node
CHECK_NODE_PORT - ockam port for checked node
CHECK_API_WORKER - API endpoint on the checked node to connect via secure channel
CHECK_PING_WORKER - worker to call through a secure channel

PROMETHEUS_PORT - port to report the metrics on

IDENTITY_IMPLEMENTATION - `sidecar` (default) or `stub` (for testing)
OCKAM_SIDECAR_HOST - identity sidecar hostname
OCKAM_SIDECAR_PORT - identity sidecar ockam port

Also improve error handling and reporting around connections and secure channels
## Checks

- [x] All commits in this Pull Request are [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) and Verified by Github.
- [x] All commits in this Pull Request follow the Ockam [commit message convention](https://github.com/build-trust/.github/blob/main/CONTRIBUTING.md#commit-messages).
- [x] I accept the Ockam Community [Code of Conduct](https://github.com/build-trust/.github/blob/main/CODE_OF_CONDUCT.md).
- [x] I have accepted the Ockam [Contributor License Agreement](https://github.com/build-trust/ockam-contributors/blob/main/CLA.md) by adding my Git/Github details in a row at the end of the [CONTRIBUTORS.csv](https://github.com/build-trust/ockam-contributors/blob/main/CONTRIBUTORS.csv) file in a separate pull request to the [build-trust/ockam-contributors](https://github.com/build-trust/ockam-contributors) repository.

### CODE REVIEW

File: implementations/elixir/ockam/ockam_healthcheck/lib/application.ex [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  Main application for Ockam Healthcheck
+  """
+
+  use Application
+
+  require Logger
+
+  @doc false
+  def start(_type, _args) do
+    # ++ healthcheck_schedule()
```
Comment (mszpakowski) [inline implementations/elixir/ockam/ockam_healthcheck/lib/application.ex]: Is that supposed to be commented out?

File: implementations/elixir/ockam/ockam_healthcheck/lib/healthcheck/metrics.ex [hunk 2]
```diff
@@ -0,0 +1,38 @@
+defmodule Ockam.Healthcheck.Metrics do
+  @moduledoc """
+  Metrics to expose healthceck results
```
Comment (mszpakowski) [inline implementations/elixir/ockam/ockam_healthcheck/lib/healthcheck/metrics.ex]: 💅 
```suggestion
  Metrics to expose healthcheck results
```

File: implementations/elixir/ockam/ockam_healthcheck/test/test_helper.exs [hunk 3]
```diff
@@ -0,0 +1,29 @@
+Application.ensure_all_started(:ockam)
+
+defmodule Test.Utils do
```
Comment (mszpakowski) [inline implementations/elixir/ockam/ockam_healthcheck/test/test_helper.exs]: Do we need this? I don't see it used anywhere, but maybe I missed something.

File: implementations/elixir/ockam/ockam_healthcheck/lib/healthcheck/metrics.ex [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+  def metrics() do
+    [
+      Metrics.counter("ockam.healthcheck.ok.count",
+        measurement: :count,
+        event_name: [:ockam, :healthcheck],
+        tags: [:node_host, :node_port, :api_worker, :ping_worker],
+        keep: fn meta ->
+          match?(%{result: :ok}, meta)
+        end
```
Comment (mszpakowski) [inline implementations/elixir/ockam/ockam_healthcheck/lib/healthcheck/metrics.ex]: Just curious why are we matching here and not explicitly emitting a `[:healthcheck, :ok]` event (same with errors one)?

**Review by mszpakowski — COMMENTED** [review]
As far as I can understand it looks good (but my understanding of all the implications is limited). Left a few mostly minor comments.

One additional suggestion that I would have is to use `HEALTHCHECK_` as env var prefix (e.g. `HEALTHCHECK_NODE_HOST`) instead of `CHECK_`, as the latter is a bit ambiguous.

File: implementations/elixir/ockam/ockam_healthcheck/lib/healthcheck/metrics.ex (hunk 4 above)
Comment (hairyhum) [reply implementations/elixir/ockam/ockam_healthcheck/lib/healthcheck/metrics.ex]: Mostly due to the prometheus reporter behaviour around tags. If we want to add `result` tag to the metrics and it's missing, the metrics is not reported. And the tag is missing in the `:ok` result.
Maybe there is a way to solve that, but I couldn't figure it out.

**Comment by hairyhum** [comment]
> One additional suggestion that I would have is to use HEALTHCHECK_ as env var prefix (e.g. HEALTHCHECK_NODE_HOST) instead of CHECK_, as the latter is a bit ambiguous.

Makes sense.

File: implementations/elixir/ockam/ockam_healthcheck/lib/healthcheck/metrics.ex (hunk 4 above)
Comment (mszpakowski) [reply implementations/elixir/ockam/ockam_healthcheck/lib/healthcheck/metrics.ex]: I'm not sure if I understand. Couldn't we do something like this:
```elixir
## Ockam.Healthcheck

  def report_check_ok(node_host, node_port, api_worker, ping_worker) do
    Telemetry.emit_event([:healthcheck, :ok],
      measurements: %{count: 1},
      metadata: %{
        node_host: node_host,
        node_port: node_port,
        api_worker: api_worker,
        ping_worker: ping_worker
      }
    )
  end

  def report_check_failed(node_host, node_port, api_worker, ping_worker, reason) do
    Telemetry.emit_event([:healthcheck, :errors],
      measurements: %{count: 1},
      metadata: %{
        node_host: node_host,
        node_port: node_port,
        api_worker: api_worker,
        ping_worker: ping_worker,
        reason: reason
      }
    )
  end

## Ockam.Healthcheck.Metrics

  def metrics() do
    [
      Metrics.counter("ockam.healthcheck.ok.count",
        measurement: :count,
        event_name: [:ockam, :healthcheck, :ok],
        tags: [:node_host, :node_port, :api_worker, :ping_worker]
      ),
      Metrics.counter("ockam.healthcheck.errors.count",
        measurement: :count,
        event_name: [:ockam, :healthcheck, :errors],
        tags: [:node_host, :node_port, :api_worker, :ping_worker, :reason],
        tag_values: fn meta ->
          case Map.fetch(meta, :reason) do
            {:ok, reason} ->
              Map.put(meta, :reason, inspect(reason))

            _other ->
              Map.put(meta, :reason, "")
          end
        end
      )
    ]
  end
```

**Review by mszpakowski — DISMISSED** [review]
LGTM, but but my understanding of all the implications (especially when it comes to changes in the `ockam` app) is limited

File: implementations/elixir/ockam/ockam_healthcheck/lib/healthcheck/metrics.ex (hunk 4 above)
Comment (hairyhum) [reply implementations/elixir/ockam/ockam_healthcheck/lib/healthcheck/metrics.ex]: Makes sense

**Comment by hairyhum** [comment]
TODO: test docker image build

**Comment by mszpakowski** [comment]
> TODO: test docker image build

I tested it locally and it builds fine, is there any other way we want to test that?

**Review by polvorin — DISMISSED** [review]

**Review by polvorin — APPROVED** [review]

**Pull request merged**